File: CheckinDemo/ViewController.m
//
//  ViewController.m
//  CheckinDemo
//
//  Created by eidan on 2017/3/29.
//  Copyright © 2017年 Amap. All rights reserved.
//

#import "ViewController.h"
#import <MAMapKit/MAMapKit.h>
#import <AMapLocationKit/AMapLocationKit.h>
#import <AMapSearchKit/AMapSearchKit.h>
#import "POITableViewCell.h"

#define AMAPLOC_DEG_TO_RAD      0.0174532925199432958f //0.017453292519943295769236907684886
#define AMAPLOC_EARTH_RADIUS    6378137.0f

@interface ViewController () <AMapLocationManagerDelegate, MAMapViewDelegate, AMapSearchDelegate>

@property (nonatomic, strong) AMapLocationManager *locationManager;
@property (nonatomic, strong) AMapSearchAPI *search;
@property (nonatomic, weak) IBOutlet MAMapView *mapView;

//data
@property (nonatomic, assign) CLLocationCoordinate2D currentGPSCoordinate;
@property (nonatomic, copy) NSArray<AMapPOI *> *POIDataArray;
@property (nonatomic, assign) int selectedIndex;

//xib views
@property (nonatomic, weak) IBOutlet UIActivityIndicatorView *loadingView;
@property (nonatomic, weak) IBOutlet UITableView *tableView;
@property (nonatomic, weak) IBOutlet UIButton *startUpdateLocationBtn;
@property (nonatomic, weak) IBOutlet UIButton *checkinBtn;


@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    //定位
    self.locationManager = [[AMapLocationManager alloc] init];
    self.locationManager.delegate = self;
    
    //搜索
    self.search = [[AMapSearchAPI alloc] init];
    self.search.delegate = self;

    //地图
    self.mapView.delegate = self;
    self.mapView.showsUserLocation = NO;
    self.mapView.rotateCameraEnabled = NO;
    self.mapView.rotateEnabled = NO;
    
    [self.tableView registerNib:[UINib nibWithNibName:@"POITableViewCell" bundle:nil] forCellReuseIdentifier:POITableViewCellIdentifier];
    UIView *view = [UIView new];
    view.backgroundColor = [UIColor clearColor];
    [self.tableView setTableFooterView:view];
    
    [self resetViewAndData];
    
}

#pragma mark - Private Action

//重设界面和数据
- (void)resetViewAndData {
    self.loadingView.hidden = YES;
    self.currentGPSCoordinate = kCLLocationCoordinate2DInvalid;  //一开始要设置为非法值，等定位成功才设置有效值
    self.selectedIndex = -1;
    self.POIDataArray = nil;
    [self.tableView reloadData];
}

//开始定位
- (void)startLocation {
    
    [self.mapView removeOverlays:self.mapView.overlays];
    [self.mapView removeAnnotations:self.mapView.annotations];
    
    self.loadingView.hidden = NO;
    self.startUpdateLocationBtn.enabled = self.checkinBtn.enabled = NO;
    
    __weak typeof(self) weakSelf = self;
    [self.locationManager requestLocationWithReGeocode:NO completionBlock:^(CLLocation *location, AMapLocationReGeocode *regeocode, NSError *error) {
        
        weakSelf.loadingView.hidden = YES;
        weakSelf.startUpdateLocationBtn.enabled = weakSelf.checkinBtn.enabled = YES;
        
        if (error) {
            NSLog(@"定位错误:{%ld - %@};", (long)error.code, error.localizedDescription);
            return;
        }
        
        weakSelf.currentGPSCoordinate = location.coordinate;
        
        //添加定位点的大头针
        MAPointAnnotation *annotation = [[MAPointAnnotation alloc] init];
        annotation.coordinate = location.coordinate;
        [weakSelf.mapView addAnnotation:annotation];
        annotation.lockedToScreen = YES;
        annotation.lockedScreenPoint = CGPointMake(weakSelf.mapView.bounds.size.width / 2, weakSelf.mapView.bounds.size.height / 2) ;
        
        //添加500米的范围圈
        MACircle *circleOverlay = [MACircle circleWithCenterCoordinate:location.coordinate radius:500];
        [weakSelf.mapView addOverlay:circleOverlay];
        
        //设置地图
        [weakSelf.mapView setZoomLevel:15.5 animated:YES];
        [weakSelf.mapView selectAnnotation:annotation animated:YES];
        [weakSelf.mapView setCenterCoordinate:location.coordinate animated:NO];
        
        //搜索POI
        [weakSelf searchAllPOIAround:location.coordinate];
    }];
}

//搜索周边POI
- (void)searchAllPOIAround:(CLLocationCoordinate2D)coordinate {
    AMapPOIAroundSearchRequest *request = [[AMapPOIAroundSearchRequest alloc] init];
    request.sortrule = 0;
    request.offset = 50;
    request.location = [AMapGeoPoint locationWithLatitude:coordinate.latitude longitude:coordinate.longitude];
    request.radius = 500;
    [self.search AMapPOIAroundSearch:request];
}

//计算两点之间的距离
- (double)distanceBetweenCoordinates:(CLLocationCoordinate2D )from and:(CLLocationCoordinate2D)to {
    
    double latitudeArc  = (from.latitude - to.latitude) * AMAPLOC_DEG_TO_RAD;
    double longitudeArc = (from.longitude - to.longitude) * AMAPLOC_DEG_TO_RAD;
    
    double latitudeH = sin(latitudeArc * 0.5);
    latitudeH *= latitudeH;
    double lontitudeH = sin(longitudeArc * 0.5);
    lontitudeH *= lontitudeH;
    
    double tmp = cos(from.latitude*AMAPLOC_DEG_TO_RAD) * cos(to.latitude*AMAPLOC_DEG_TO_RAD);
    return AMAPLOC_EARTH_RADIUS * 2.0 * asin(sqrt(latitudeH + tmp*lontitudeH));
}

#pragma mark - xib click

//开始定位
- (IBAction)startUpdateLocation:(id)sender {
    [self resetViewAndData];
    [self startLocation];
}

//签到
- (IBAction)checkin:(id)sender {
    if (CLLocationCoordinate2DIsValid(self.currentGPSCoordinate)) {  //定位成功了，就可以签到
        [self.mapView setZoomLevel:19 animated:YES];
        UIAlertView *alertView = [[UIAlertView alloc] initWithTitle:@"签到成功" message:nil delegate:nil cancelButtonTitle:@"确定" otherButtonTitles:nil, nil];
        [alertView show];
    } else {
        UIAlertView *alertView = [[UIAlertView alloc] initWithTitle:@"位置在火星？~~签到失败" message:nil delegate:nil cancelButtonTitle:@"确定" otherButtonTitles:nil, nil];
        [alertView show];
    }
}

#pragma -mark UITableView Delegate and DataSource

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.POIDataArray.count;
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    return 62;
}

- (UITableViewCell*)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    
    POITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:POITableViewCellIdentifier forIndexPath:indexPath];
    cell.accessibilityIdentifier = [NSString stringWithFormat:@"POITableViewCell_%ld",(long)indexPath.row];  //for UITest
    
    AMapPOI *POI = self.POIDataArray[indexPath.row];
    
    cell.nameLabel.text = POI.name;
    cell.infoLabel.text = POI.address;
    cell.selectedImageView.hidden = indexPath.row == self.selectedIndex ? NO : YES;
    
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath{
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    
    self.selectedIndex = (int)indexPath.row;
    [self.tableView reloadData];
    
    AMapPOI *POI = self.POIDataArray[indexPath.row];
    [self.mapView setCenterCoordinate:CLLocationCoordinate2DMake(POI.location.latitude, POI.location.longitude) animated:YES];
    [self.mapView setZoomLevel:19 animated:YES];
}

#pragma mark - AMapLocationManagerDelegate

- (void)amapLocationManager:(AMapLocationManager *)manager doRequireLocationAuth:(CLLocationManager*)locationManager {
    [locationManager requestAlwaysAuthorization];
}

#pragma mark - AMapSearchDelegate

- (void)onPOISearchDone:(AMapPOISearchBaseRequest *)request response:(AMapPOISearchResponse *)response{
    self.POIDataArray = response.pois;
    [self.tableView reloadData];
}

#pragma mark - MAMapViewDelegate

- (void)mapViewRequireLocationAuth:(CLLocationManager *)locationManager {
    [locationManager requestAlwaysAuthorization];
}

- (void)mapView:(MAMapView *)mapView mapDidMoveByUser:(BOOL)wasUserAction {
    if (!wasUserAction) {
        return;
    }
    if (!CLLocationCoordinate2DIsValid(self.currentGPSCoordinate)) {  //非法的时候需返回
        return;
    }
    double dis = [self distanceBetweenCoordinates:mapView.centerCoordinate and:self.currentGPSCoordinate];
    if (dis > 500 ) {
        UIAlertView *alertView = [[UIAlertView alloc] initWithTitle:@"调整距离不可超过500米" message:nil delegate:self cancelButtonTitle:@"确定" otherButtonTitles:nil, nil];
        [alertView show];
    }
}

- (void)alertView:(UIAlertView *)alertView clickedButtonAtIndex:(NSInteger)buttonIndex {
    [self.mapView setZoomLevel:15.5 animated:YES];
    [self.mapView setCenterCoordinate:self.currentGPSCoordinate animated:YES];
}



- (MAAnnotationView *)mapView:(MAMapView *)mapView viewForAnnotation:(id<MAAnnotation>)annotation {
    if ([annotation isKindOfClass:[MAPointAnnotation class]]) {
        static NSString *pointReuseIndetifier = @"pointReuseIndetifier";
        
        MAPinAnnotationView *annotationView = (MAPinAnnotationView *)[mapView dequeueReusableAnnotationViewWithIdentifier:pointReuseIndetifier];
        if (annotationView == nil) {
            annotationView = [[MAPinAnnotationView alloc] initWithAnnotation:annotation reuseIdentifier:pointReuseIndetifier];
        }
        
        annotationView.canShowCallout = NO;
        annotationView.animatesDrop = NO;
        annotationView.draggable = NO;
        
        return annotationView;
    }
    
    return nil;
}

- (MAOverlayRenderer *)mapView:(MAMapView *)mapView rendererForOverlay:(id <MAOverlay>)overlay {
    
    if ([overlay isKindOfClass:[MACircle class]]) {
        MACircleRenderer *circleRenderer = [[MACircleRenderer alloc] initWithCircle:overlay];
        circleRenderer.lineWidth = 2.0f;
        circleRenderer.strokeColor = [UIColor orangeColor];
        circleRenderer.fillColor = [UIColor clearColor];
        
        return circleRenderer;
    }
    
    return nil;
}


- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}


@end

File: CheckinDemo-swift/ViewController.swift
//
//  ViewController.swift
//  CheckinDemo-swift
//
//  Created by eidan on 2017/3/30.
//  Copyright © 2017年 Amap. All rights reserved.
//

import UIKit

class ViewController: UIViewController ,AMapLocationManagerDelegate, MAMapViewDelegate, AMapSearchDelegate, UITableViewDelegate, UITableViewDataSource{
    
    let AMAPLOC_DEG_TO_RAD = 0.0174532925199432958
    let AMAPLOC_EARTH_RADIUS = 6378137.0
    
    let POITableViewCellIdentifier = "POITableViewCellIdentifier"
    
    var locationManager: AMapLocationManager!   //定位
    var search: AMapSearchAPI!                  //地图内的搜索API类
    @IBOutlet weak var mapView: MAMapView!      //地图
    
    //data
    var currentGPSCoordinate: CLLocationCoordinate2D?
    var POIDataArray = [AMapPOI]()
    var selectedIndex = -1
    
    //xib views
    @IBOutlet weak var loadingView: UIActivityIndicatorView!
    @IBOutlet weak var tableView: UITableView!
    @IBOutlet weak var startUpdateLocationBtn: UIButton!
    @IBOutlet weak var checkinBtn: UIButton!

    override func viewDidLoad() {
        super.viewDidLoad()
        
        //定位
        self.locationManager = AMapLocationManager.init()
        self.locationManager.delegate = self
        
        //Search
        self.search = AMapSearchAPI.init()
        self.search.delegate = self
        
        //map
        self.mapView.delegate = self
        self.mapView.showsUserLocation = false
        self.mapView.isRotateCameraEnabled = false
        self.mapView.isRotateEnabled = false
        
        self.tableView.register(UINib.init(nibName: "POITableViewCellSwift", bundle: nil), forCellReuseIdentifier: POITableViewCellIdentifier)
        let view: UIView = UIView.init()
        view.backgroundColor = UIColor.clear
        self.tableView.tableFooterView = view
        
        //搜索POI
        self.resetViewAndData()
        
        // Do any additional setup after loading the view, typically from a nib.
    }
    
    // MARK: - Private Action
    
    //重设界面和数据
    func resetViewAndData() {
        self.loadingView.isHidden = true
        self.currentGPSCoordinate = kCLLocationCoordinate2DInvalid
        self.selectedIndex = -1
        self.POIDataArray =  [AMapPOI]()
        self.tableView.reloadData()
    }
    
    //开始定位
    func startLocation() {
        self.mapView.removeOverlays(self.mapView.overlays)
        self.mapView.removeAnnotations(self.mapView.annotations)
        
        self.loadingView.isHidden = false
        self.startUpdateLocationBtn.isEnabled = false
        self.checkinBtn.isEnabled = false
        
        self.locationManager.requestLocation(withReGeocode: false) { [weak self] (location: CLLocation?, regeocode: AMapLocationReGeocode?, error: Error?) in
            
            self?.loadingView.isHidden = true
            self?.startUpdateLocationBtn.isEnabled = true
            self?.checkinBtn.isEnabled = true
            
            if let error = error {
                
                let error = error as NSError
                print("locError:{\(error.code) - \(error.localizedDescription)};")
                
                return
            }
            
            self?.currentGPSCoordinate = location?.coordinate
            
            //添加定位点的大头针
            let annotation: MAPointAnnotation = MAPointAnnotation.init()
            annotation.coordinate = (location?.coordinate)!
            self?.mapView.addAnnotation(annotation)
            annotation.isLockedToScreen = true
            annotation.lockedScreenPoint = CGPoint.init(x: (self?.mapView.bounds.size.width)! / 2, y: (self?.mapView.bounds.size.height)! / 2)
            
            //添加500米的范围圈
            let circleOverlay: MACircle = MACircle.init(center: (location?.coordinate)!, radius: 500)
            self?.mapView.add(circleOverlay)
            
            //设置地图
            self?.mapView.setZoomLevel(15.5, animated: true)
            self?.mapView.selectAnnotation(annotation, animated: true)
            self?.mapView.centerCoordinate = (location?.coordinate)!
            
            self?.searchAllPOIAround(coordinate: (location?.coordinate)!)
            
        }
        
    }
    
    //搜索周边POI
    func searchAllPOIAround(coordinate: CLLocationCoordinate2D) {
        let request: AMapPOIAroundSearchRequest = AMapPOIAroundSearchRequest.init()
        request.sortrule = 0
        request.offset = 50
        request.location = AMapGeoPoint.location(withLatitude: CGFloat(coordinate.latitude), longitude: CGFloat(coordinate.longitude))
        request.radius = 500
        self.search.aMapPOIAroundSearch(request)
    }
    
    //计算两点之间的距离
    func distanceBetweenCoordinates(from: CLLocationCoordinate2D, to: CLLocationCoordinate2D) -> Double {
        
        let latitudeArc  = (from.latitude - to.latitude) * AMAPLOC_DEG_TO_RAD;
        let longitudeArc = (from.longitude - to.longitude) * AMAPLOC_DEG_TO_RAD;
        
        var latitudeH = sin(latitudeArc * 0.5);
        latitudeH *= latitudeH;
        var lontitudeH = sin(longitudeArc * 0.5);
        lontitudeH *= lontitudeH;
        
        let tmp = cos(from.latitude * AMAPLOC_DEG_TO_RAD) * cos(to.latitude * AMAPLOC_DEG_TO_RAD);
        return AMAPLOC_EARTH_RADIUS * 2.0 * asin(sqrt(latitudeH + tmp * lontitudeH));
    }
    
    // MARK: - xib btn click
    
    //开始定位
    @IBAction func startUpdateLocation(_ sender: Any) {
        self.resetViewAndData()
        self.startLocation()
    }
    
    //签到
    @IBAction func checkin(_ sender: UIButton) {
        if CLLocationCoordinate2DIsValid(self.currentGPSCoordinate!) {
            self.mapView.setZoomLevel(19.0, animated: true)
            let alertView: UIAlertView = UIAlertView.init(title: "签到成功", message: nil, delegate: nil, cancelButtonTitle: "确定")
            alertView.show()
        } else {
            let alertView: UIAlertView = UIAlertView.init(title: "位置在火星？~~签到失败", message: nil, delegate: nil, cancelButtonTitle: "确定")
            alertView.show()
        }
    }
    
    // MARK: - UITableView Delegate and DataSource
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return self.POIDataArray.count
    }
    
    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return 62
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        
        let cell: POITableViewCellSwift = tableView.dequeueReusableCell(withIdentifier: POITableViewCellIdentifier, for: indexPath) as! POITableViewCellSwift
        
        let POI: AMapPOI = self.POIDataArray[indexPath.row];
        
        cell.nameLabel.text = POI.name
        cell.infoLabel.text = POI.address
        cell.selectedImageView.isHidden = indexPath.row == self.selectedIndex ? false : true;
        
        return cell
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
        
        self.selectedIndex = indexPath.row
        self.tableView.reloadData()
        
        let POI: AMapPOI = self.POIDataArray[indexPath.row]
        self.mapView.setCenter(CLLocationCoordinate2DMake(CLLocationDegrees(POI.location.latitude), CLLocationDegrees(POI.location.longitude)), animated: true)
        self.mapView.setZoomLevel(19.0, animated: true)
    }
    
    // MARK: - AMapLocationManagerDelegate
    
    func amapLocationManager(_ manager: AMapLocationManager!, doRequireLocationAuth locationManager: CLLocationManager!) {
        locationManager.requestAlwaysAuthorization()
    }
    
    // MARK: - AMapSearchDelegate
    
    func onPOISearchDone(_ request: AMapPOISearchBaseRequest!, response: AMapPOISearchResponse!) {
        self.POIDataArray = response.pois
        self.tableView.reloadData()
    }
    
    // MARK: - MapViewDelegate
    //  地图移动结束后调用此接口 wasUserAction 标识是否是用户动作
    func mapView(_ mapView: MAMapView!, mapDidMoveByUser wasUserAction: Bool) {
        if wasUserAction == false {
            return;
        }
        
        if CLLocationCoordinate2DIsValid(self.currentGPSCoordinate!) == false {
            return
        }
        
        let dis = self.distanceBetweenCoordinates(from: mapView.centerCoordinate, to: self.currentGPSCoordinate!)
        
        if dis > 500 {
            self.mapView.setZoomLevel(15.5, animated: true)
            self.mapView.setCenter(self.currentGPSCoordinate!, animated: true)
            let alertView: UIAlertView = UIAlertView.init(title: "调整距离不可超过500米", message: nil, delegate: nil, cancelButtonTitle: "确定")
            alertView.show()
        }
        
    }
    
    func mapView(_ mapView: MAMapView!, viewFor annotation: MAAnnotation!) -> MAAnnotationView! {
        
        if annotation is MAPointAnnotation {
            
            let pointReuseIndetifier: String = "pointReuseIndetifier"
            
            var annotationView = mapView.dequeueReusableAnnotationView(withIdentifier: pointReuseIndetifier) as? MAPinAnnotationView
            
            if annotationView == nil {
                annotationView = MAPinAnnotationView.init(annotation: annotation, reuseIdentifier: pointReuseIndetifier)
                
                annotationView?.canShowCallout = false
                annotationView?.isDraggable = false
                annotationView?.animatesDrop = false
            }
            
            return annotationView
        }
        
        return nil
    }
    
    func mapView(_ mapView: MAMapView!, rendererFor overlay: MAOverlay!) -> MAOverlayRenderer! {
        
        if overlay.isKind(of: MACircle.self) {
            let circleRenderer: MACircleRenderer = MACircleRenderer(circle: overlay as! MACircle!)
            circleRenderer.lineWidth = 2
            circleRenderer.strokeColor = UIColor.orange
            circleRenderer.fillColor = UIColor.clear
            return circleRenderer
        }
        return nil
    }


    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }


}

